crates/networking/rpc/based/versioned_message.rs

```diff
@@ -0,0 +1,245 @@
+    pub signature: Signature,
+    pub message: VersionedMessage,
+}
+
+impl SignedMessage {
+    fn parse(params: &Option<Vec<serde_json::Value>>) -> Result<Self, RpcErr> {
+        tracing::info!("Parsing based message");
+
+        let Some(params) = params else {
+            return Err(RpcErr::InvalidBasedMessage(
+                "Failed to parse based request into SignedMessage, params were expected but none were found".to_string(),
+            ));
+        };
+
+        serde_json::from_value(
+            params
+                .first()
+                .ok_or(RpcErr::InvalidBasedMessage(
+                    "Failed to parse based request into SignedMessage, message not found"
+                        .to_string(),
+                ))?
+                .clone(),
+        )
+        .map_err(|e| {
+            RpcErr::InvalidBasedMessage(format!(
+                "Failed to parse based request into SignedMessage: {e}"
+            ))
+        })
+    }
+
+    pub fn call_env(req: &RpcRequest, context: RpcApiContext) -> Result<serde_json::Value, RpcErr> {
+        let request = Self::parse(&req.params)?;
+        match &request.message {
+            VersionedMessage::EnvV0(env) => {
+                request.check_signature(&context.gateway_pubkey)?;
+                env.handle(context)
+            }
+            other_based_message => Err(RpcErr::InvalidBasedMessage(format!(
+                "Failed to handle Env message, found {other_based_message:?} instead of Env"
+            ))),
+        }
+    }
+
+    pub fn call_new_frag(
+        req: &RpcRequest,
+        context: RpcApiContext,
+    ) -> Result<serde_json::Value, RpcErr> {
+        let request = Self::parse(&req.params)?;
+        match &request.message {
+            VersionedMessage::FragV0(frag) => {
+                request.check_signature(&context.gateway_pubkey)?;
+                frag.handle(context)
+            }
+            other_based_message => Err(RpcErr::InvalidBasedMessage(format!(
+                "Failed to handle Frag message, found {other_based_message:?} instead of Frag"
+            ))),
+        }
+    }
+
+    pub fn call_seal_frag(
+        req: &RpcRequest,
+        context: RpcApiContext,
+    ) -> Result<serde_json::Value, RpcErr> {
+        let request = Self::parse(&req.params)?;
+        match &request.message {
+            VersionedMessage::SealV0(seal) => {
+                request.check_signature(&context.gateway_pubkey)?;
+                seal.handle(context)
+            }
+            other_based_message => Err(RpcErr::InvalidBasedMessage(format!(
+                "Failed to handle Seal message, found {other_based_message:?} instead of Seal"
+            ))),
+        }
+    }
+
+    fn check_signature(&self, expected: &Public) -> Result<(), RpcErr> {
+        let message = libsecp256k1::Message::parse(&self.message.tree_hash_root().0);
+        let signature = libsecp256k1::Signature::parse_standard_slice(&self.signature[..64])
```

Comment:
Let's use `get` instead of brackets slicing here.

Reply:
Done in [5f52088](https://github.com/lambdaclass/ethrex/pull/2274/commits/5f52088405d3950903eb309ae9d1c4a5394969d9). Also found more places to do this.